Added MG brand vehicles based on SAIC interface

kscholty opened the pull request:
Implementation of MG branded vehicles based on the new SAIC ISmart interface.

This implementation is based on the fabulous work of [tisof (Markus Heberling)](https://github.com/tisoft).
A Discussion and a link to a demo implementation in Java can be found at
https://github.com/orgs/SAIC-iSmart-API/discussions/18

I tested it with my MG4. Seems to work. Unfortunately the API tends to deliver invalid or incomplete results when the car is not charging or driving (in sleep mode or just woke up)
I tried to invalidate caches when appropriate in order to find a compromise between number of queries and good results.


Would love to get feedback how it behaves in the real world.
Comments on the code welcome as well

kscholty committed:
This is an implementation of the SAIC interface and the MG car brand.
This is untested.
Currently the required fields are not 100% clear.

Comments welcome

kscholty committed:
This is an implementation of the SAIC interface and the MG car brand.
This is untested.
Currently the required fields are not 100% clear.

Comments welcome

kscholty committed:
This is an implementation of the SAIC interface and the MG car brand.
This is untested.
Currently the required fields are not 100% clear.

Comments welcome

kscholty committed:
Merge branch 'feature/SAIC' of https://github.com/kscholty/evcc into feature/SAIC

kscholty committed:
Formatting

kscholty committed:
Formatting

andig commented on vehicle/saic/requests/sendRequest.go:
```diff
@@ -0,0 +1,100 @@
+	"time"
+)
+
+func SendRequest(
```
If we need this for all requests, a transport decorator would be nice to use.

kscholty replied:
I refactored the communication. Now some of the headers are added by a decorator.
Not sure that this is what you intended. Only the constant headers can be added by a decorator. All the dynamic ones are also used in the body (or for the encryption) so it doesn't make sense to delegate that to a decorator.
However, I tried to stick to the already implemented logic of using local transports for the API and the identity.

andig replied:
> Only the constant headers can be added by a decorator.

oauth2 decorates the token. We could do the same and make the entire request encryption a decorator. That may be more a topic of "nicer code" though.

kscholty replied:
Ok. Didn't know that decorators can be used to modify the body of a rquest.

andig commented:
Is there a reference implementation for the new interface?

kscholty commented:
This implementation is based on the  fabulous work of [tisof (Markus Heberling)](https://github.com/tisoft).
A Discussion and a link to a demo implementation in Java can be found at
[https://github.com/orgs/SAIC-iSmart-API/discussions/18](https://github.com/orgs/SAIC-iSmart-API/discussions/18)

kscholty committed:
Formatting fixes

kscholty committed:
Formatting fixes

kscholty committed:
More formatting

kscholty committed:
More formatting

kscholty committed:
And even more formatting...

kscholty committed:
And even more formatting...

kscholty committed:
Refactored the communication logic.
Plit up the sendong routine to give the caller more access to the result
and used a transport deocartor to add some of the headers.

kscholty committed:
Refactored the communication logic.
Plit up the sendong routine to give the caller more access to the result
and used a transport deocartor to add some of the headers.

andig commented on vehicle/saic/identity.go:
```diff
@@ -0,0 +1,118 @@
+	for i := 0; i < 64; i++ {
+		deviceId[i] = byte(rand.Intn(255))
+	}
+	v.deviceId = hex.EncodeToString(deviceId[:]) + "###com.saicmotor.europecar"
```
lo.RandomString

kscholty replied:
Good hint.

kscholty committed:
Fixes after testing.
Mainly improved error handling.

kscholty committed:
Fixes after testing.
Mainly improved error handling.

kscholty committed:
Merge branch 'master' into feature/SAIC

kscholty committed:
Merge branch 'feature/SAIC' of https://github.com/kscholty/evcc into feature/SAIC

andig commented:
@kscholty bist Du soweit fertig? Dann müsste das jetzt grundsätzlich an den evcc Stil angepasst werden. Ich könnte mir das anschauen, brauche dafür aber Zugangsdaten zum testen. Bitte an info@evcc.io. Danke!

kscholty committed:
Fixed odometer value

kscholty commented:
> @kscholty bist Du soweit fertig? Dann müsste das jetzt grundsätzlich an den evcc Stil angepasst werden. Ich könnte mir das anschauen, brauche dafür aber Zugangsdaten zum testen. Bitte an [info@evcc.io](mailto:info@evcc.io). Danke!

PN ist unterwegs

andig committed:
Refactor

andig committed:
Simplify

kscholty committed:
Merge branch 'evcc-io:master' into feature/SAIC

kscholty committed:
Merge branch 'evcc-io:master' into feature/SAIC

kscholty commented:
@andig Was machen wir den hier jetzt? Ich kann mit meinem Auto gar nicht mehr testen. Bekommen weder hier noch in der offiziellen App irgendwelche Daten. Keine Ahnung, warum.
Ich denke, wir sollten das anderen mal zum Testen zur Verfügung stellen, um zu sehen, wie es bei anderen Fahrzeugen funktioniert.

kscholty committed:
Merge branch 'evcc-io:master' into feature/SAIC

kscholty committed:
Added background check for data.

kscholty commented:
Ich habe im letzen commit die API so erweitert, dass sie bei einem Fehlschlag der Abfrage ("Try again later") im Hintergrund weiter versucht, Daten zu bekommen. Das verbessert die Erfolgsquote deutlich. LEider werden dann immer wieder mal falsche Daten geliefert. Ich habe dafür keine Erklärung.

Es ist jedenfalls so, dass das Fahrzug zuverlässig Daten liefert, wenn es geladen wird. Dabei habe ich bisher keine Probleme feststellen können.

kscholty committed:
Merge branch 'evcc-io:master' into feature/SAIC

kscholty committed:
Merge branch 'evcc-io:master' into feature/SAIC

kscholty committed:
Added SocLimiter interface to provider

kscholty commented:
Hi @andig,

wie ist denn dein Plan hier weiter zu machen? Für mich ist das so funktional. Ich lade mein Auto seit  über einer Woche damit problemlos.
Was kann ich tun, das für dich akzeptabler zu machen?

Gruß
kscholty

kscholty committed:
Merge branch 'evcc-io:master' into feature/SAIC

andig commented:
Funktioniert es denn jetzt zuverlässig ohne Aussetzer und Fakedaten?

kscholty commented:
> Funktioniert es denn jetzt zuverlässig ohne Aussetzer und Fakedaten?

Es ist so, dass mein Fahrzeug immer wieder Kommunikationsprobleme hat. Offensichtlich wird jeder Aufruf ans Fahrzeug weitergeleitet, was manchmal einfach nicht funktioniert. Ich habe eine Woche lange jeden Tag geladen und es hat problemlos funktioniert, zumindest immer dann, wenn das Fahrzeug geladen wurde.  Im Moment gibts wieder Probleme.

Die Probleme hängen aber nicht mit der Implementierung hier zusammen, sondern mit dem Fahrzeug oder den Servern von MG, denn die offizielle App verhält sich identisch. Wenn evcc keine Daten bekommt, dann auch die App nicht und umgekehrt.
Ich bin absolut sicher, dass die Implentierung korrekt ist und wir sie anderen MG Fahrern zum Testen zur Verfügung stellen sollten.

kscholty committed:
Merge branch 'evcc-io:master' into feature/SAIC

andig commented:
Na dann mal rein damit 👍🏻

premultiply commented:
@kscholty Danke für deine Arbeit!

tma202 commented:
Wie kann ich es testen ? Ist es im Nightly ?

kscholty commented:
> Wie kann ich es testen ? Ist es im Nightly ?

Ist auf jeden Fall im master.

premultiply commented:
...und ich habe mir sagen lassen: Es funktioniert.

tma202 commented:
Danke,benötige ich trotzdem das mqtt Gateway ?Dachte ich könnte es nun ganz normal einrichten wie z.B. einen ID 3 ?GrußToniGesendet mit der WEB.DE iPad AppAm 04.03.24 um 14:28 schrieb kscholty
                
            
            
                
                    Von: "kscholty" ***@***.***>Datum: 4. März 2024An: "evcc-io/evcc" ***@***.***>Cc: "tma202" ***@***.***>,"Comment" ***@***.***>Betreff: Re: [evcc-io/evcc] Added MG brand vehicles based on SAIC interface (PR #12003)
                
                

Wie kann ich es testen ? Ist es im Nightly ?

Ist auf jeden Fall im master.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>

tma202 commented:
Habe gerade gelesen das, das Template MG heiß, werde mir dann anschauen. 
kannst du mir ein Beispiel angeben ? Dankeschön !

andig commented:
@tma202 nutz bitte

    evcc configure

Das ist selbsterklärend.
